Sources/SwiftGRPCNIO/HTTP1ToRawGRPCServerCodec.swift

```diff
@@ -43,29 +50,64 @@ public final class HTTP1ToRawGRPCServerCodec {
+  // are encouraged to unwrap them using guard-else statements. These don't work cleanly
+  // with structs, since the guard-else would create a new copy of the struct, which
+  // would then have to be re-assigned into the class variable for the changes to take effect.
+  // By force unwrapping, we avoid those reassignments, and the code is a bit cleaner.
+
+  // Buffer to store binary encoded protos as they're being received.
+  private var binaryRequestBuffer: NIO.ByteBuffer!
+
+  // Buffers to store text encoded protos. Only used when content-type is application/grpc-web-text.
+  // TODO(kaipi): Extract all gRPC Web processing logic into an independent handler only added on
+  // the HTTP1.1 pipeline, as it's starting to get in the way of readability.
+  private var requestTextBuffer: NIO.ByteBuffer!
+  private var responseTextBuffer: NIO.ByteBuffer!
 }
 
 extension HTTP1ToRawGRPCServerCodec: ChannelInboundHandler {
   public typealias InboundIn = HTTPServerRequestPart
   public typealias InboundOut = RawGRPCServerRequestPart
-  
+
   public func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
     switch self.unwrapInboundIn(data) {
     case .head(let requestHead):
       guard case .expectingHeaders = state
         else { preconditionFailure("received headers while in state \(state)") }
 
       state = .expectingCompressedFlag
-      buffer = ctx.channel.allocator.buffer(capacity: 5)
-
+      binaryRequestBuffer = ctx.channel.allocator.buffer(capacity: 5)
+      if let contentTypeHeader = requestHead.headers["content-type"].first {
+        contentType = ContentType(rawValue: contentTypeHeader)
+      }
+      if contentType == .text {
+        requestTextBuffer = ctx.channel.allocator.buffer(capacity: 0)
+      }
       ctx.fireChannelRead(self.wrapInboundOut(.head(requestHead)))
 
     case .body(var body):
-      guard var buffer = buffer
-        else { preconditionFailure("buffer not initialized") }
+      precondition(binaryRequestBuffer != nil, "buffer not initialized")
       assert(state.expectingBody, "received body while in state \(state)")
-      buffer.write(buffer: &body)
+
+      // If the contentType is text, then decode the incoming bytes as base64 encoded, and append
+      // it to the binary buffer. If the request is chunked, this section will process the text
+      // in the biggest chunk that is multiple of 4, leaving the unread bytes in the textBuffer
+      // where it will expect a new incoming chunk.
+      if contentType == .text {
+        precondition(requestTextBuffer != nil)
+        requestTextBuffer.write(buffer: &body)
+        // Read in chunks of 4 bytes , as base64 encoded strings will always be multiples of 4.
+        let readyBytes = requestTextBuffer.readableBytes - (requestTextBuffer.readableBytes % 4)
+        if let base64Encoded = requestTextBuffer.readString(length: readyBytes),
+          let decodedData = Data(base64Encoded: base64Encoded) {
```

Comment:
Should we cancel the call if the data can not be decoded as base64?

Reply:
closed the connection, but not sure how to handle it properly from the server side, or what to send to the client, since at this point the headers have already been sent. should we wait until the request has been fully decoded before starting to send headers? (out of scope for this change though)

Reply:
Not sure, either. Thanks for adding the `FIXME`!